TheaterSchedule.BLL/Services/ImageService.cs

```diff
@@ -1,16 +1,188 @@
 using TheaterSchedule.BLL.Interfaces;
+using TheaterSchedule.DAL.Interfaces;
+using TheaterSchedule.BLL.DTO;
+
 namespace TheaterSchedule.BLL.Services
 {
-   public class ImageService: IImageService
+    public class ImageService : IImageService
     {
+        private IImageRepository imageRepository;
+        private IMemoryCache cache;
+
+        private int cacheExpirationMinutes = 20;
+
+        public ImageService(IImageRepository imgRepository, IMemoryCache memoryCache)
+        {
+            imageRepository = imgRepository;
+            cache = memoryCache;
+        }
+
         public string ImageToBase64(byte[] image)
         {
             string base64Image = Convert.ToBase64String(image);
             return base64Image;
         }
-    }
 
+        public string GeMimeTypeFromImageByteArray(byte[] byteArray)
+        {
+            try
+            {
+                using (MemoryStream stream = new MemoryStream(byteArray))
+                {
+                    using (Image image = Image.FromStream(stream))
+                    {
+                        return ImageCodecInfo.GetImageEncoders().First(codec => codec.FormatID == image.RawFormat.Guid).MimeType;
+                    }
+                }
+            }
+            catch (ArgumentException)
+            {
+                return "text/plain";
+            }
+        }
+
+        public async Task<ImageBase64DTO> LoadPerformanceMainImageBase64Async(int id)
+        {
+            ImageBase64DTO imageDTO = null;
+            byte[] image = await TryAddPerformanceImageToCacheAsync(id);
+
+            if (image != null)
+            {
+                await Task.Run(() =>
+                {
+                    imageDTO = new ImageBase64DTO()
+                    {
+                        Image = ImageToBase64(image),
+                        ImageFormat = GeMimeTypeFromImageByteArray(image)
+                    };
+                });
+            }
+
+            return imageDTO;
+        }
+
+        public async Task<ImageBytesDTO> LoadPerformanceMainImageBytesAsync(int id)
+        {
+            ImageBytesDTO imageDTO = null;
+            byte[] image = await TryAddPerformanceImageToCacheAsync(id);
+
+            if (image != null)
+            {
+                await Task.Run(() =>
+                {
+                    imageDTO = new ImageBytesDTO()
+                    {
+                        Image = image,
+                        ImageFormat = GeMimeTypeFromImageByteArray(image)
+                    };
+                });
+            }
+
+            return imageDTO;
+        }
+
+        public async Task<List<ImageBase64DTO>> LoadPerformanceGalleryBase64Async(int id)
+        {
+            List<ImageBase64DTO> imagesDTO = new List<ImageBase64DTO>();
+            IEnumerable<byte[]> images = await TryAddGalleryImagesToCacheAsync(id);
+
+            await Task.Run(() =>
+            {
+                foreach (var image in images)

```

Comment:
images.Select(image => new ImageBase64DTO()
                    {
                        Image = ImageToBase64(image),
                        ImageFormat = GeMimeTypeFromImageByteArray(image)
                    })
Use linq Select - it easier for understanding and more compact